### ISSUE TITLE
Problems building from Matlab R2018a with "Visual Studio 15 2017" 

### ISSUE THREAD

**Issue by plebelshawca** [jlblancoc/suitesparse-metis-for-windows#75]
Are there any updates that apply to this?
I use windows 10.
I get a lot of the following when building from Matlab R2018a with "Visual Studio 15 2017" :

..\..\metis-5.1.0\GKlib\gk_types.h(22): error C2061: syntax error: identifier 'gk_idx_t'
..\..\metis-5.1.0\GKlib\gk_types.h(22): error C2059: syntax error: ';'
..\..\metis-5.1.0\GKlib\gk_struct.h(24): error C2061: syntax error: identifier 'ssize_t'
..\..\metis-5.1.0\GKlib\gk_struct.h(24): error C2059: syntax error: '}'
..\..\metis-5.1.0\GKlib\gk_struct.h(25): error C2061: syntax error: identifier 'ssize_t'
..\..\metis-5.1.0\GKlib\gk_struct.h(25): error C2059: syntax error: '}'
..\..\metis-5.1.0\GKlib\gk_struct.h(26): error C2061: syntax error: identifier 'ssize_t'
..\..\metis-5.1.0\GKlib\gk_struct.h(26): error C2059: syntax error: '}'
..\..\metis-5.1.0\GKlib\gk_struct.h(27): error C2061: syntax error: identifier 'ssize_t'
..\..\metis-5.1.0\GKlib\gk_struct.h(27): error C2059: syntax error: '}'
..\..\metis-5.1.0\GKlib\gk_struct.h(28): error C2061: syntax error: identifier 'ssize_t'
..\..\metis-5.1.0\GKlib\gk_struct.h(28): error C2061: syntax error: identifier 'val'
..\..\metis-5.1.0\GKlib\gk_struct.h(28): error C2059: syntax error: ';'
..\..\metis-5.1.0\GKlib\gk_struct.h(28): error C2059: syntax error: '}'

Yet I can use make/cmake to create METIS.sln Visual Studio Project that I seem to be able to compile
I also seem to have similar problems compiling directly from Cygwin.
(Somewhere I saw a comment that this is untested).

**Comment by jlblancoc** [comment]
I think I don't fully understand what you mean with: 

>  building from Matlab R2018a

This project is aimed at providing a CMake way to generate VS solutions, so anything else (as Cygwin, MinGW,etc.) is untested and probably incompatible...

**Comment by pflashan** [comment]
Any chance this is problem #23 again? I was getting similar output in VS2017 when building Metis. I applied the same fix, namely to edit metis\GKLib\gkarch.h to only define rint(x) if _MSC_VER < 1800. The relevant section of gkarch.h looks like this:
```
#ifdef __MSC__
/* MSC does not have rint() function */
#if (_MSC_VER < 1800)
#define rint(x) ((int)((x)+0.5))  
#endif
```
Probably worth a shot - it fixed my errors.

**Comment by NeroBurner** [comment]
in cmake there is specific code to check if `rint` exists to prevent redefinition
https://github.com/jlblancoc/suitesparse-metis-for-windows/blob/e03725414c23ba3cf3c4f5b14b684595f75eb785/metis/CMakeLists.txt#L20-L24

**Comment by jlblancoc** [comment]
@plebelshawca I just rebuilt everything in `master` using MSVC 2017 and it builds cleanly (well... hundreds of warnings, but that's an upstream issue ;-). 

Are you using the version from `master` or the last release (1.4.0)?

**Comment by plebelshawca** [comment]
Jose,

Thanks very much for your assistance on this.

I’m sorry in that I wasn’t too clear about exactly what my issue was.

I have also been away for a bit, and so didn’t reply to you very quickly.

 

I was able to use the vs solution to build he entire metis-5.1.0 package.

I did encounter the problem identified by pflashan referring to problem #23, but this, I resolved easily.

 

I was not able to run the matlab script: “SuiteSparse_install.m”

It failed on the metis compile with the errors shown. This occurred in the compile of the (first step) UMFPACK build.

 

My attempts to compile the entire suitesparse package at once also failed in a similar manner.

I plan to attack it in a piecewise manner and see if I have any better success.

 

I’ll post my progress and watch for any advice in the meantime.

Again, Thanks.

 

plebelshawca
